source.go

```diff
@@ -1,98 +1,197 @@
+	}
+
 	return nil
 }
 
 func (s *Source) Read(ctx context.Context) (sdk.Record, error) {
-	// Read returns a new Record and is supposed to block until there is either
-	// a new record or the context gets cancelled. It can also return the error
-	// ErrBackoffRetry to signal to the SDK it should call Read again with a
-	// backoff retry.
-	// If Read receives a cancelled context or the context is cancelled while
-	// Read is running it must stop retrieving new records from the source
-	// system and start returning records that have already been buffered. If
-	// there are no buffered records left Read must return the context error to
-	// signal a graceful stop. If Read returns ErrBackoffRetry while the context
-	// is cancelled it will also signal that there are no records left and Read
-	// won't be called again.
-	// After Read returns an error the function won't be called again (except if
-	// the error is ErrBackoffRetry, as mentioned above).
-	// Read can be called concurrently with Ack.
-	return sdk.Record{}, nil
+	msg, err := s.consumer.Receive(ctx)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("failed to receive message: %w", err)
+	}
+
+	msgID := msg.ID()
+
+	position := Position{
+		MessageID:        msgID.Serialize(),
+		SubscriptionName: s.config.SubscriptionName,
+	}
+	sdkPos := position.ToSDKPosition()
+
+	metadata := sdk.Metadata{"pulsar.topic": msg.Topic()}
+	metadata.SetCreatedAt(msg.EventTime())
+
+	key := sdk.RawData(msg.Key())
+	payload := sdk.RawData(msg.Payload())
+
+	newRecord := sdk.Util.Source.NewRecordCreate(sdkPos, metadata, key, payload)
+
+	sdk.Logger(ctx).Debug().Msg("received message")
+
+	return newRecord, nil
 }
 
 func (s *Source) Ack(ctx context.Context, position sdk.Position) error {
-	// Ack signals to the implementation that the record with the supplied
-	// position was successfully processed. This method might be called after
-	// the context of Read is already cancelled, since there might be
-	// outstanding acks that need to be delivered. When Teardown is called it is
-	// guaranteed there won't be any more calls to Ack.
-	// Ack can be called concurrently with Read.
-	return nil
+	parsed, err := parsePosition(position)
+	if err != nil {
+		return err
+	}
+
+	msgID, err := pulsar.DeserializeMessageID(parsed.MessageID)
+	if err != nil {
+		return fmt.Errorf("failed to deserialize message ID: %w", err)
+	}
+
+	sdk.Logger(ctx).Debug().Str("MessageID", msgID.String()).Msg("acked message")
+
+	return s.consumer.AckID(msgID)
 }
 
 func (s *Source) Teardown(ctx context.Context) error {
-	// Teardown signals to the plugin that there will be no more calls to any
-	// other function. After Teardown returns, the plugin should be ready for a
-	// graceful shutdown.
+	if s.consumer != nil {
+		s.consumer.Close()
+	}
+
+	if s.client != nil {
+		s.client.Close()
+	}
+
+	sdk.Logger(ctx).Debug().Msg("source teardown complete")
+
 	return nil
 }
+
+type Position struct {
```

Comment:
we can have this struct and the two related methods in a separate file, I think it would look better, and would be easier to use in tests if needed

Reply:
I thought about it, but I'm really against 20 line files (even in golang, as new files don't add much complexity compared to, say, js). Position was only used in `source.go`, so I just left it there. You'll see that in the rabbitmq connector I change my mind and put it in an `utils.go` file. :shrug: 

If we ever need tests for Position, we add that refactor, shouldn't be much problematic.